lib/state/settings/actions.ts

```diff
@@ -103,3 +103,16 @@ export const toggleAutoHideMenuBar = () => (dispatch, getState) => {
     autoHideMenuBar: newValue,
   });
 };
+
+export const setLanguageDetection = languageDetectionEnabled => ({
+  type: 'setLanguageDetection',
+  languageDetectionEnabled,
+});
+
+export const toggleLanguageDetection = () => (dispatch, getState) => {
+  const {
+    settings: { languageDetectionEnabled },
+  } = getState();
+
+  dispatch(setLanguageDetection(!languageDetectionEnabled));
+};
```

Comment:
hey! fun fact - we don't ned to create imperative _thunk_-style action creators for toggle-actions like this. reducers are great at keeping things simpler and then our actions remain a basic data object representing the action our users want to take.

```ts
// state/action-types
export type ToggleLanguageDetection = Action<'TOGGLE_LANGUAGE_DETECTION'>;

// …/settings/actions.ts
import * as A from '../action-types';

export const toggleLanguageDetection: A.ActionCreator<A.ToggleLanguageDetection> = () => ( {
	type: 'TOGGLE_LANGUAGE_DETECTION',
} );

// …/settings/reducer.ts
const languageDetection: A.Reducer<boolean> = (state = false, action) => {
	switch (action.type) {
		case 'SET_LANGUAGE_DETECTION':
			return action.languageDetectionEnabled;

		case 'TOGGLE_LANGUAGE_DETECTION':
			return !state;

		default:
			return state;
	}
}
```

Note here that we never need to base our intentions/actions off of the current app state - all we have to do is explain what we're trying to accomplish and then the reducers govern their own data type. It gets murkier when we mash together multiple data values into a single reducer, which is where `combineReducers` can help (you can see how things were prior to #1216 - sometimes the boiler plate is valuable for other reasons, here in keeping separate data types isolated from each other and making it easier to reason about changes from the user's perspective).

It may also be the case that when we're working this way we don't need the `SET_` action at all. This is the css with the edit/preview mode toggle, because the app starts in one mode and all that matters from the user's perspective is the intention to toggle the mode, not necessary to _set_ it.

Reply:
Yes, that makes sense. And I see the benefits of the `combineReducers` approach better now. Thanks for the tip!